cuda_core/cuda/core/experimental/_memory/_memory_pool.pyx

```diff
@@ -0,0 +1,438 @@
+            If the buffer is deallocated without an explicit stream, the allocation stream
+            is used.
+        """
+        stream = Stream_accept(stream) if stream is not None else default_stream()
+        _MP_deallocate(self, <uintptr_t>ptr, size, <Stream> stream)
+
+    @property
+    def attributes(self) -> _MemPoolAttributes:
+        """Memory pool attributes."""
+        if self._attributes is None:
+            ref = weakref.ref(self)
+            self._attributes = _MemPoolAttributes._init(ref)
+        return self._attributes
+
+    @property
+    def device_id(self) -> int:
+        """The associated device ordinal."""
+        return self._dev_id
+
+    @property
+    def handle(self) -> driver.CUmemoryPool:
+        """Handle to the underlying memory pool."""
+        return driver.CUmemoryPool(<uintptr_t>(self._handle))
+
+    @property
+    def is_handle_owned(self) -> bool:
+        """Whether the memory resource handle is owned. If False, ``close`` has no effect."""
+        return self._mempool_owned
+
+    @property
+    def peer_accessible_by(self):
+        """
+        Get or set the devices that can access allocations from this memory
+        pool. Access can be modified at any time and affects all allocations
+        from this memory pool.
+
+        Returns a tuple of sorted device IDs that currently have peer access to
+        allocations from this memory pool.
+
+        When setting, accepts a sequence of Device objects or device IDs.
+        Setting to an empty sequence revokes all peer access.
+
+        Examples
+        --------
+        >>> dmr = DeviceMemoryResource(0)
+        >>> dmr.peer_accessible_by = [1]  # Grant access to device 1
+        >>> assert dmr.peer_accessible_by == (1,)
+        >>> dmr.peer_accessible_by = []  # Revoke access
+        """
+        return self._peer_accessible_by
+
+    @peer_accessible_by.setter
+    def peer_accessible_by(self, devices):
+        """Set which devices can access this memory pool."""
+        from .._device import Device
+
+        # Convert all devices to device IDs
+        cdef set[int] target_ids = {Device(dev).device_id for dev in devices}
+        target_ids.discard(self._dev_id)  # exclude this device from peer access list
+        this_dev = Device(self._dev_id)
+        cdef list bad = [dev for dev in target_ids if not this_dev.can_access_peer(dev)]
+        if bad:
+            raise ValueError(f"Device {self._dev_id} cannot access peer(s): {', '.join(map(str, bad))}")
+        cdef set[int] cur_ids = set(self._peer_accessible_by)
+        cdef set[int] to_add = target_ids - cur_ids
+        cdef set[int] to_rm = cur_ids - target_ids
+        cdef size_t count = len(to_add) + len(to_rm) # transaction size
+        cdef cydriver.CUmemAccessDesc* access_desc = NULL
+        cdef size_t i = 0
+
+        if count > 0:
+            access_desc = <cydriver.CUmemAccessDesc*>PyMem_Malloc(count * sizeof(cydriver.CUmemAccessDesc))
```

Comment:
Is there any way we could use something like a `std::vector` here or something else that has proper RAII semantics instead of having to manually allocate and free? I.E.
```
access_desc = cdef vector[CUmemAccessDesc] access_descs
access_desc.reserve(count)
```

It looks like you have things handled here in the error cases, but it always makes me wary of doing manual memory management if we don't have to.